compiler/pipes/final-check.cpp

```diff
@@ -527,6 +543,46 @@ void FinalCheckPass::check_instanceof(VertexAdaptor<op_instanceof> instanceof_ve
   kphp_error(instanceof_var_type->class_type(), fmt_format("left operand of 'instanceof' should be an instance, but passed {}", instanceof_var_type->as_human_readable()));
 }
 
+static void check_indexing_violation(std::string_view allowed_types_string, const std::vector<PrimitiveType> &allowed_types, std::string_view what_indexing,
+                                     const TypeData &key_type) noexcept {
+  kphp_error(vk::contains(allowed_types, key_type.ptype()),
+             fmt_format("Only {} types are allowed for {}{}indexing, but {} type is passed", allowed_types_string, what_indexing,
+                        what_indexing.empty() ? "" : " ", key_type.as_human_readable()));
+}
+
+static void check_indexing_type(PrimitiveType indexed_type, const TypeData &key_type) noexcept {
+  if (key_type.get_real_ptype() == tp_bool) {
+    return;
+  }
+
+  if (indexed_type == tp_tuple || indexed_type == tp_string) {
+    static const std::vector<PrimitiveType> allowed_types{tp_int, tp_float, tp_mixed, tp_future};
+    check_indexing_violation("int, float and future", allowed_types, ptype_name(indexed_type), key_type);
+  } else {
+    static const std::vector<PrimitiveType> allowed_types{tp_int, tp_float, tp_string, tp_mixed, tp_future};
```

Comment:
Is tp_float reasonable for arrays indexing?

Reply:
To me not, but there are bunch of compilation issues in production. After discussion we decided that better leave as is.